### DISCUSSION TITLE
For each block not recognizing list in data

### DISCUSSION THREAD

**Discussion by domenicr** [postmanlabs/postman-flows#176] [Defect]
The for each block is not recognizing or detecting the list within the data. Does the data need to be formatted in a certain way?

![image](https://user-images.githubusercontent.com/8070525/146802585-dd953d6c-b59d-4bdf-8bf6-970dffe4309a.png)

**Comment by hydramind** [comment]
Same issue for me. Anyone have any ideas?

**Comment by samay-v** [comment]
Hey, @domenicr 
Can you share a sample of the response body that is referred here?

**Comment by domenicr** [comment]
Hi Samay,

I wanted to generate the file and it seems to be working now,
See attached file.

Thanks,
Domenic

**Comment by alexei-moneyr** [comment]
As I see, postman team changed the block (foreach) and without any documentation. 
Now it doesnt recognize the list as before. How do I use it now?!
What is the new structure? and why did you changed it? I need to updated a lot of flows now (

**Reply by alexei-moneyr** [reply]
I see why my list is felt.
Concentrate now takes only lists. 
How do I combine two variables (json) into list? Concentrate only applies on durables, not on variables anymore. 
all my flows are not working anymore because of this change (

Please help us understand your moves. 
Thanks

**Reply by alexei-moneyr** [reply]
found a way, but.... it's not nice at all....
have to add another step

![image](https://user-images.githubusercontent.com/82156498/148698150-0032f5fc-e6d4-4e17-b53d-606a05ca6858.png)

**Reply by saswatds** [reply]
Hey @alexei-moneyr I think the problem is with the type-inference system. 
We have made two improvements over the for-each block
1. Added support for full expressions
2.  Added type-checking
I guess a combination of incorrect type inference and additional type-checking is what is causing the issue. 

The `Create Variable` block was a stop-gap solution because we did not have a way to create data in flows. With the new `Create Data` block, you should be able to create all data structures. We would wish all users to start using that instead. But I hear you.

Let's say we implement this behaviour to Concatenate block, will it solve your use-case? 
1. two lists of same type -> list
2. two lists of different type -> tuple
3. One list and one value of matching type -> list 
4. One list and one value of different type -> tuple
5. Two values of same type -> list
6. Two values of different type -> tuple

The only downside that I see with this is that it becomes very confusing to understand what exactly is happening in the flow.

**Reply by alexei-moneyr** [reply]
Using data type is ok, resolved the issue, but I have to recreate all variable blocks (I have 20-30 of them in different flows) with 12 fields in each.
This is painful. ((

![image](https://user-images.githubusercontent.com/82156498/148881830-6f146d93-7aec-4297-951c-56cc6132a69a.png)

![image](https://user-images.githubusercontent.com/82156498/148881859-e1859407-214f-4397-9036-b2fa0c72f1c9.png)

data type list  -> record with needed fields.

Please stop changing datatypes.

**Comment by tomalazarov** [comment]
@samay-v is that a bug? I stumbled on the same issue. This is something that worked some days ago... we are talking about simple foreach... but after last update you've messed up something with the data types. I just want to iterate through an array of accounts but instead got an error thrown "Expected list, got record".  Can you please help me? Here are some screenshots:
<img width="942" alt="Screenshot 2022-01-10 at 18 35 21" src="https://user-images.githubusercontent.com/48013428/148805298-4ea8d8ba-490f-4ab5-9a5c-101bc6d8db5f.png">
<img width="1169" alt="Screenshot 2022-01-10 at 18 43 57" src="https://user-images.githubusercontent.com/48013428/148805308-c0925b19-4352-40f7-b1df-c6bdbbacdbff.png">

**Reply by alexei-moneyr** [reply]
try my solution. may help you. 
but it's ugly one

**Reply by saswatds** [reply]
@tomalazarov could you help me with a sample of your actual data? For some reason, the type inference system is considering your array to be a tuple. You can send it to help@postman.com with sensitive information replaced.

**Reply by tomalazarov** [reply]
10x... I understood how things work and now the forEach block recognises that the "Accounts" is a list and I can iterate through it's elements. It's like first I should run the Account List request and then append the forEach block, then it recognise it as a list. I don't know what I was doing but somehow I wasn't  giving the forEach block a proper response and it didn't understand what exactly I'm passing to it (the forEach block was taking the Accounts as a record)... but anyways now everything is fine and fixed and working properly!

**Comment by tomalazarov** [comment]
There's definitely an issue with the For Each block. I don't know why, when and how exactly it occurs but this is what happens - I append the forEach block and everything is fine > I leave this particular flow with the forEach and do some other stuff > When I come back to the flow with forEach and the block is highlighted in red and not recognising the response it gets as a list anymore... but takes it as a null or record > In order to make it work again I have to remove it and set it again. This is the way I'm overpassing this issue but it's only temporary solution and I have to do it over and over again and it's very annoying!

Please, fix it ;(
@saswatds 

**Edit: I believe the issue is even bigger** because it applies also for Send Request block and the "Assign values to variables" functionality. I'll just give an example because it's easier to explain - Imagine you have GET Travel Plans request and after it Delete Travel Plan; so in order to delete a travel plan I just take the "requestID" of the first record (travel plan) from the list with travel plans. Everything is working fine and done just for minutes. When I come back tomorrow to that flow and boom - the Delete Travel Plan block is not anymore recognising the requestID variable and **an error saying "Expected bool, number, string, timestamp, got null" is displayed.** And once again I have to remove the variable and add it again to make it work... for sometime until it's broken again :(

**Reply by saswatds** [reply]
@tomalazarov I just realized what the issue is here. We have this feature in flows that tries to figure out the schema of the response body after the request had been sent. Now this cache is not persisted across restart and architecturally it was getting impossible to maintain it as it was bringing coupling between runtime and compile time . 
An alternative for that that was to define an `Example` for the request which flows could always use to infer the schema of the response body. If you can see that the request is in a warning state and asking to add an example.

In the last release we added a full-fledged type-system implementation to do type validation as part of the compilation step. Now when the flow is reopened all the previously cached types are getting lost and compilation error is not letting anything to run.

I will try figure-out a way in which user won't need to define examples (Maybe attach an OpenAPI3 spec or something). But mean while if you could just save an Example with the response body for the request, it should start working.

You can use this doc for reference to add an example https://learning.postman.com/docs/sending-requests/examples/

**Reply by tomalazarov** [reply]
Many thanks for the quick answer. I'll try that solution and give feedback later!

**Comment by avenak** [comment]
Same issue here. My send request returns this:
`[
  {
    "id": 58
  },
  {
    "id": 77
  }
]`
Hence, the body of this response is a List (of Records), _not_ a Record. No matter what I do, the For Each block insists that this is a Record. If the type system cannot infer that this is a list (the fact that it's JSON and there are square brackets - big hint), then the entire Flows feature has a massive, fundamental issue. Such a shame - I really hoped this would work but it doesn't, so flow deleted and back to the drawing board.

**Reply by avenak** [reply]
And yes, my request does have an example. And I am using the latest desktop version of Postman (v9.15.2). I just tried the via the web interface - same issue.
